### DISCUSSION TITLE
Esp32 psram + ili9341  How to do?

### DISCUSSION THREAD

**Discussion by straga** [lvgl-micropython/lvgl_micropython#106] [Board Specific]
Display work tested with old version mpdisplay, but firmware from that repo without new SPI C driver use Micropython Native SPI all works.

```
from lcd_bus_2 import SPIBus
from ili9341 import ILI9341
from machine import Pin, SPI
from xpt2046 import Touch

display_bus = SPIBus(
    dc=5,
    host=1,
    mosi=13,
    miso=12,
    sclk=14,
    cs=15,
    freq=40_000_000,
    wp=-1,
    hd=-1,
    quad_spi=False,
    tx_only=True,
    cmd_bits=8,
    param_bits=8,
    dc_low_on_data=False,
    sio_mode=False,
    lsb_first=False,
    cs_high_active=False,
    spi_mode=0,
)

display_drv = ILI9341(
    display_bus,
    width=240,
    height=320,
    colstart=0,
    rowstart=0,
    rotation=270,
    mirrored=False,
    color_depth=16,
    bgr=True,
    reverse_bytes_in_word=True,
    invert=False,
    brightness=1.0,
    backlight_pin=None,
    backlight_on_high=True,
    reset_pin=4,
    reset_high=False,
    power_pin=22,
    power_on_high=True,
)

```

If build firmware with New SPI driver and not use anymore mpdisplay.

Look like that. Not errors. But not render. Just Activate Dispaly and Horizontal Line.

![telegram-cloud-photo-size-4-5947337383840170376-y](https://github.com/user-attachments/assets/46c193ad-cacf-4729-b45c-c7b157aeec7f)

```
import sys
import lvgl as lv
import lcd_bus

import ili9341
from machine import SPI, Pin

spi_bus = SPI.Bus(
    host=1,
    mosi=13,
    miso=12,
    sck=14
)

display_bus = lcd_bus.SPIBus(
    spi_bus=spi_bus,
    dc=5,
    cs=15,
    freq=40000000,
)

# fb1 = display_bus.allocate_framebuffer(_BUFFER_SIZE, lcd_bus.MEMORY_SPIRAM)#lcd_bus.MEMORY_INTERNAL | lcd_bus.MEMORY_DMA)
# fb2 = display_bus.allocate_framebuffer(_BUFFER_SIZE, lcd_bus.MEMORY_SPIRAM)#lcd_bus.MEMORY_INTERNAL | lcd_bus.MEMORY_DMA)

import lvgl as lv
fb_size = 240*320*lv.color_format_get_size(lv.COLOR_FORMAT.RGB565)
fb1 = display_bus.allocate_framebuffer(fb_size, lcd_bus.MEMORY_SPIRAM)
fb2 = display_bus.allocate_framebuffer(fb_size, lcd_bus.MEMORY_SPIRAM)

display = ili9341.ILI9341(
    data_bus=display_bus,
    display_width=240,
    display_height=320,
    frame_buffer1=fb1,
    frame_buffer2=fb2,
    reset_pin=4,
    reset_state=False,
    backlight_pin=22,
    color_space=lv.COLOR_FORMAT.RGB565,
    # color_byte_order=ili9341.BYTE_ORDER_BGR,
    rgb565_byte_swap=True
)

display.init()
display.set_rotation(lv.DISPLAY_ROTATION._90)
display.set_backlight(50)
display.invert_colors()

lv.init()
scrn = lv.obj()
scrn.set_style_bg_color(lv.color_hex(0xffb6c1), 0)

slider = lv.slider(scrn)
slider.set_size(300, 50)
slider.center()

label = lv.label(scrn)
label.set_text('HELLO WORLD!')
label.align(lv.ALIGN.CENTER, 0, -50)
```

**Comment by kdschlosser** [comment]
give this a try and see if it works. The frame buffers you don't have to mess about with, let the display driver take care of setting those up.

```python
import sys
import lvgl as lv
import lcd_bus

import ili9341
from machine import SPI, Pin

spi_bus = SPI.Bus(
    host=1,
    mosi=13,
    miso=12,
    sck=14
)

display_bus = lcd_bus.SPIBus(
    spi_bus=spi_bus,
    dc=5,
    cs=15,
    freq=40000000,
)

import lvgl as lv


display = ili9341.ILI9341(
    data_bus=display_bus,
    display_width=240,
    display_height=320,
    reset_pin=4,
    # comment the next line if you still don't get something on the display
    reset_state=ili9341.STATE_LOW,
    backlight_pin=22,
    color_space=lv.COLOR_FORMAT.RGB565,
    # color_byte_order=ili9341.BYTE_ORDER_BGR,
    rgb565_byte_swap=True
)

display.set_power(True)
display.init()
display.set_rotation(lv.DISPLAY_ROTATION._90)

# setting this to anything but 100 or 0 when you don't have the 
# backlight state set to PWM isn't going to do anything
display.set_backlight(100)

display.invert_colors()

scrn = lv.screen_active()
scrn.set_style_bg_color(lv.color_hex(0xffb6c1), 0)

slider = lv.slider(scrn)
slider.set_size(300, 50)
slider.center()

label = lv.label(scrn)
label.set_text('HELLO WORLD!')
label.align(lv.ALIGN.CENTER, 0, -50)

import task_handler
th = task_handler.TaskHandler()
```

**Reply by straga** [reply]
OK. Thanks. Now works.
![telegram-cloud-photo-size-4-5947337383840170429-y](https://github.com/user-attachments/assets/ea1eea01-3049-43e1-9880-592839224cf5)

**Reply by fabse-hack** [reply]
nice!!!

**Comment by straga** [comment]
Just need do some extra by hand. Now tested with isp-idf 5.2.2 with mpy 1.24 preview + lvgl 9.1. Work LCD and Touch: ESP32 psram with and ILI9341 2.8” display and XPT2046. Also works asyncio task from push button on LCD.

**Reply by kdschlosser** [reply]
You don't need to do anything with the submodules. It is all taken care of for you through the build script. It will clone the proper IDF version, MicroPython version and LVGL version. It is not something you have to mess about with. Just clone the repo and run the build command adding the `submodules` and `mpy_cross`  parameter to the command the first time you run the build command. If you run the build command again you do nomt have to provide those parameters in the build command.

**Comment by kdschlosser** [comment]
It works and It works quite well and is actually really easy to get up and running if you don't try and do things like how they are done in the "official" lvgl micropython binding.

This library's sole goal was to make things easier to do, making it easy to update LVGL and MicroPython versions, adding support for more displays and touch drivers and the last thing was to fix issues like not being able to have the touch and display on the same SPI bus. and not being able to use the SDCard reader if it is on the same bus as the touch or the display. All of those things have been fixed.

**Comment by straga** [comment]
it's a very cool project that allows you to use lvgl. Very thanks about that.
Some time need build firmware only with (lvgl,lcd_spi and ... ) without change inside Board and without frozen file in firmware, already have own board definition. Also esp-idf env already have to.

**Comment by fabse-hack** [comment]
nice! i hope i can run my setting soon with lvgl and micropython

**Comment by fabse-hack** [comment]
how you get the code to your controller??

`python.exe -m mpremote connect COM6 cp main.py :` ??????

is here the whole lvgl binding similar to the micropython ?

**Reply by straga** [reply]
@fabse-hack 
I am use ftp for file upload and UART or Telnet for repl.

Right setting for me.
```
display = ili9341.ILI9341(
    data_bus=display_bus,
    display_width=240,
    display_height=320,
    reset_pin=4,
    # comment the next line if you still don't get something on the display
    reset_state=ili9341.STATE_LOW,
    #power_pin=22,
    backlight_pin=22,
    backlight_on_state=ili9341.STATE_PWM,
    color_space=lv.COLOR_FORMAT.RGB565,
    color_byte_order=ili9341.BYTE_ORDER_BGR,
    rgb565_byte_swap=True
)
#display.invert_colors()

```

![telegram-cloud-photo-size-4-5956036058414170863-y](https://github.com/user-attachments/assets/ac508c41-988d-4d1e-ab02-e47d0be92d89)

**Reply by kdschlosser** [reply]
Looks like ya got it running. cool..

**Comment by androiddrew** [comment]
@straga Could you provide the complete working example you have? I am working on an ESP32-S3 and trying to get this all setup myself.

**Reply by androiddrew** [reply]
![IMG_0785](https://github.com/user-attachments/assets/6ad069fb-9bb4-4aa7-9f76-1342be041ca6)

Ok I got something to work on my ESP32-S3. @kdschlosser I am really struggling how to connect the python objects to the the LVGL c API. I know this is a skills issue but if you could help connect the dots in some fashion I think I'll be able to take this and build out a bunch of examples that I think other people will be able to follow.

**Reply by kdschlosser** [reply]
Here is some code you can look at to get a basic idea. This is for an older version of LVGL so some of the names for things will be different but it should give you a general idea of how the organization of things is done.

https://github.com/kdschlosser/mp_lvgl_widgets

**Comment by straga** [comment]
some example.

```
import lv_config
import lvgl as lv
import asyncio
from asyncio import sleep, create_task, Loop

scr = lv.screen_active()
scr.set_style_bg_color(lv.color_hex(0x808080), 0) # gray
# scr.set_style_bg_color(lv.color_hex(0xffb6c1), 0) # pink
# import task_handler
# th = task_handler.TaskHandler()


async def btn_reset_task(obj=None, event=-1):

    print(f"Resetting... Taks rom: {obj}, event: {event}")
    scr.clean()


    rst_spinner = lv.spinner(scr)
    rst_spinner.set_size(100, 100)
    rst_spinner.center()
    rst_spinner.set_anim_params(10000, 60)
    rst_spinner.set_style_bg_color(lv.color_hex(0x808080), 0)  # gray

    rst_label = lv.label(scr)
    rst_label.center()
    rst_label.set_text("Progress...")

    print("Resetting...")
    await asyncio.sleep(5)
    print("Resetting... Done")

    # import machine
    # machine.reset()

    scr.clean()
    btn_reset_gen()

def btn_reset_task_act(arg1, arg2):
    print(f"Resetting... arg1: {arg1}, arg2: {arg2}")
    # print("Resetting...")

def btn_reset_gen():
    # Create a style for the button label

    style_btn_label = lv.style_t()
    style_btn_label.init()
    style_btn_label.set_text_color(lv.color_hex(0xFFFFFF))  # Set text color to white

    btn_reset = lv.button(scr)
    btn_reset.align(lv.ALIGN.TOP_MID, 0, 280)
    btn_reset.add_event_cb(lambda e: create_task(btn_reset_task()), lv.EVENT.CLICKED, None)
    label_reset = lv.label(btn_reset)
    label_reset.set_text('Reset Task')
    label_reset.add_style(style_btn_label, 0)  # Apply the style to the label

    btn_reset = lv.button(scr)
    btn_reset.align(lv.ALIGN.TOP_MID, 0, 10)
    btn_reset.add_event_cb(lambda e: print("Resetting..."), lv.EVENT.CLICKED, None)
    label_reset = lv.label(btn_reset)
    label_reset.set_text('Reset Print')
    label_reset.add_style(style_btn_label, 0)  # Apply the style to the label


btn_reset_gen()

import lv_utils
if not lv_utils.event_loop.is_running():
    eventloop = lv_utils.event_loop(asynchronous=True)

loop = asyncio.get_event_loop()
loop.run_forever()
```
